pgjdbc/src/main/java/org/postgresql/jdbc/SQLDataReader.java

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright (c) 2024, PostgreSQL Global Development Group
+ * See the LICENSE file in the project root for more information.
+ */
+
+package org.postgresql.jdbc;
+
+import static java.lang.Character.isWhitespace;
+
+import org.postgresql.core.BaseConnection;
+
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+import java.sql.SQLData;
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.List;
+
+public class SQLDataReader {
+  public @Nullable  <T> T read(@Nullable String value, Class<T> type, BaseConnection connection, TimestampUtils timestampUtils) throws SQLException {
+    if (value == null) {
+      return null;
+    }
+    SQLData data;
+    try {
+      data = (SQLData) type.getConstructor().newInstance();
+    } catch (Exception ex) {
+      throw new SQLException(String.format("An accessible no-arg constructor is required for type [%s]", type), ex);
+    }
+
+    data.readSQL(new PgSQLInput(parse(value, '(', ')'), connection, timestampUtils), data.getSQLTypeName());
+
+    return type.cast(data);
+  }
+
+  public List<@Nullable String> parseArray(String value) {
+    return parse(value, '{', '}');
+  }
+
+  private List<@Nullable String> parse(String value, char begin, char end) {
+    List<@Nullable String> values = new ArrayList<>();
+
+    int len = value.length();
+    StringBuilder builder = null;
+
+    int lastDelimIdx = -1;
+
+    scan: for (int charIdx = 0; charIdx < len; ++charIdx) {
+      char ch = value.charAt(charIdx);
+      if (ch == begin) {
+        lastDelimIdx = charIdx;
+      } else if (ch == end) {
+        addTextElement(builder, lastDelimIdx, charIdx, values);
+        break scan;
+      } else if (ch == '"') {
+        builder = new StringBuilder();
+        charIdx = readString(value, charIdx, builder);
+      } else if (ch == ',') {
+        addTextElement(builder, lastDelimIdx, charIdx, values);
+        builder = null;
+        lastDelimIdx = charIdx;
+      } else {
+        //
+        // ignore any whitespace we encounter
+        //
+        if (isWhitespace(ch)) {
```

Comment:
Check the comment in the conversation about this

Reply:
Sorry all, I have been swamped at my work. I will get back to this. Thank you for continuing to review this work.

Reply:
@crowmagnumb work does get in the way. Thanks for working on this.